lib/plug/upload.ex

```diff
@@ -155,22 +156,29 @@ defmodule Plug.Upload do
     Process.flag(:trap_exit, true)
     tmp = Enum.find_value(@temp_env_vars, "/tmp", &System.get_env/1)
     cwd = Path.join(File.cwd!(), "tmp")
-    :ets.new(@table, [:named_table, :public, :set])
+
+    :ets.new(@pid_table, [:named_table, :public, :set])
+    :ets.new(@path_table, [:named_table, :public, :duplicate_bag])
+
     {:ok, [tmp, cwd]}
   end
 
   @impl true
-  def handle_call(:upload, {pid, _ref}, dirs) do
+  def handle_call({:monitor, pid}, _from, dirs) do
     Process.monitor(pid)
     {:reply, {:ok, dirs}, dirs}
   end
 
   @impl true
   def handle_info({:DOWN, _ref, :process, pid, _reason}, state) do
-    case :ets.lookup(@table, pid) do
-      [{pid, _tmp, paths}] ->
-        :ets.delete(@table, pid)
-        delete_paths(paths)
+    case :ets.lookup(@pid_table, pid) do
+      [{pid, _tmp}] ->
+        :ets.delete(@pid_table, pid)
+
+        @path_table
+        |> :ets.lookup(pid)
+        |> Enum.map(&elem(&1, 1))
+        |> delete_paths()
```

Comment:
```suggestion
        @path_table
        |> :ets.lookup(pid)
        |> Enum.each(&delete_path/1)
```

And then in terminate you can do:

    folder = fn entry, :ok -> delete_path(entry) end